### ISSUE TITLE
[macOS] some UI elements (e.g. list, tree views) seem to have more space between lines now

### ISSUE THREAD

**Issue by martinlippert** [eclipse-platform/eclipse.platform.ui#1674]
Some UI elements seem to have a lot more space between lines on macOS since Eclipse 2024-03 builds, for example the package explorer, dialogs showing lists (switch to different perspective), etc. It looks like a very general issue (maybe even in SWT, not sure). Not sure if this was an intended change, but I would prefer the old behavior to save screen real estate.

Eclipse 2024-03-M2 on macOS Sonoma 14.3.1
Plain vanilla EPP package download, empty workspace.

**Comment by martinlippert** [comment]
This is how it looks like for 2024-03 M2 on macOS:

<img width="386" alt="package-explorer-2024-03-m2-on-macOS" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/0cb0e330-6600-42cb-92aa-42b8cf58beea">

vs. the overall look using 2023-12 on macOS:

<img width="386" alt="package-explorer-2023-12-on-macOS" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/49606e05-9f92-4c9d-89f3-1bc5fdc55466">

**Comment by merks** [comment]
Have you tried any of the SWT snippets to see if the problem is like that there too?

**Comment by BeckerWdf** [comment]
I think https://github.com/eclipse-platform/eclipse.platform.swt/pull/771 is the case for this. And I had the impression that is was by intention.

**Comment by Phillipus** [comment]
The underlying cause is the macOS SDK version that the Eclipse launcher is built with. In older versions of Eclipse the launcher was built using SDK 10. Later versions are built using SDK 11 and 13. This dictates how the UI elements are displayed. Taller tree and table items are part of SDK 11 and greater. https://github.com/eclipse-platform/eclipse.platform.swt/pull/771 is a fix to keep things consistent.

**Comment by martinlippert** [comment]
> The underlying cause is the macOS SDK version that the Eclipse launcher is built with. In older versions of Eclipse the launcher was built using SDK 10. Later versions are built using SDK 11 and 13. This dictates how the UI elements are displayed. Taller tree and table items are part of SDK 11 and greater. [eclipse-platform/eclipse.platform.swt#771](https://github.com/eclipse-platform/eclipse.platform.swt/pull/771) is a fix to keep things consistent.

I tested this using I20240219-0600, which has `org.eclipse.swt.cocoa.macosx.aarch64_3.125.0.v20240217-1555`. This build still has the described issue. I am not sure whether the fix that got merged in https://github.com/eclipse-platform/eclipse.platform.swt/pull/771 is already included in that I-build though.

**Comment by Phillipus** [comment]
> This build still has the described issue.

Not sure what you mean?

To be clear, since Eclipse 4.30 the launcher has been built using SDK 13 so taller row heights are the norm. https://github.com/eclipse-platform/eclipse.platform.swt/pull/771 was merged on Jan 17 2024. All that patch does is to ensure that row heights stay as they should when a font is set.

**Comment by martinlippert** [comment]
> > This build still has the described issue.
> 
> Not sure what you mean?
> 
> To be clear, since Eclipse 4.30 the launcher has been built using SDK 13 so taller row heights are the norm. [eclipse-platform/eclipse.platform.swt#771](https://github.com/eclipse-platform/eclipse.platform.swt/pull/771) was merged on Jan 17 2024. All that patch does is to ensure that row heights stay as they should when a font is set.

I got the impression that the mentioned PR fixes the issue that I illustrated in the screenshots in https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-1940598475, that is why I thought I should re-test the latest I-build and report back here... ;-)

The screenshot above from using Eclipse 2023-12 (4.30) looks like the old behavior though (smaller row height), so it looks like something has changed for 4.31 builds that causes this change.

**Comment by Phillipus** [comment]
Since Eclipse 4.30 all row heights on Mac are taller by default. Before https://github.com/eclipse-platform/eclipse.platform.swt/pull/771 changing the font of a Table/Tree/List incorrectly added padding of 1 pixel to the row height (wrong). It now adds 8 pixels to match the increased row height. You are seeing smaller row heights in Eclipse 2023-12 because the patch is not present in that version.

The correct row height is the larger one, because of the changes in the SDK.

**Comment by Phillipus** [comment]
I'll try to make this clearer...

Please take a look at https://github.com/eclipse-platform/eclipse.platform.swt/issues/677

- Since Eclipse 4.30 the x86_64 launcher is built using SDK 13 (has always been the case for Silicon which was built with SDK 11 and now 13) so all row heights are now natively taller (25 pixels instead of 18)
- In Eclipse 4.30 you might see a smaller row height because a padding of 1 pixel was being added to the row instead of 8 pixels when the font was set on the tree/table/list (as in your screenshot)
- This was fixed in https://github.com/eclipse-platform/eclipse.platform.swt/pull/771 so that all row heights remain consistent at 25 pixels high either before or after changing the font on the control

TL;DR - taller row heights is the new normal.

**Comment by Phillipus** [comment]
If you want smaller row height you could use `-Dorg.eclipse.swt.internal.carbon.smallFonts` in your eclipse.ini file.

**Comment by martinlippert** [comment]
> `-Dorg.eclipse.swt.internal.carbon.smallFonts`

The above screenshot when running on Eclipse 2024-03 M2 is with `-Dorg.eclipse.swt.internal.carbon.smallFonts` on the `eclipse.ini` file.

**Comment by BeckerWdf** [comment]
> > `-Dorg.eclipse.swt.internal.carbon.smallFonts`
> 
> The above screenshot when running on Eclipse 2024-03 M2 is with `-Dorg.eclipse.swt.internal.carbon.smallFonts` on the `eclipse.ini` file.

If you remove `-Dorg.eclipse.swt.internal.carbon.smallFonts` you should see that you even see less entries on the same screen space in project explorer.

**Comment by martinlippert** [comment]
Okay, here is a comparison (from the users point of view), first with the smallFonts setting:

A) Eclipse 2023-12 (including `-Dorg.eclipse.swt.internal.carbon.smallFonts`)
<img width="357" alt="eclipse-2023-12-with-small-fonts" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/8d84e63b-4aae-47d6-ada8-0404a870e371">

B) Eclipse 2024-03 I-Build (including `-Dorg.eclipse.swt.internal.carbon.smallFonts`)
<img width="357" alt="eclipse-2024-03-i-build-with-small-fonts" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/e7fb7f92-60c2-4acc-870a-4e44d1066e2b">

The screenshots for both without `-Dorg.eclipse.swt.internal.carbon.smallFonts` look the same, so posted here just for the reference:

C) Eclipse 2023-12 (excluding `-Dorg.eclipse.swt.internal.carbon.smallFonts`)
<img width="357" alt="eclipse-2023-12-without-small-fonts" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/40cee289-fd28-4370-b833-b96945df3559">

D) Eclipse 2024-03 I-Build (excluding `-Dorg.eclipse.swt.internal.carbon.smallFonts`)
<img width="357" alt="eclipse-2023-12-without-small-fonts" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/815b38ff-b5dc-4eb3-826a-33a19144da76">

---

I would actually prefer to continue to have the more compressed look of A) for all my views, lists, tress, etc., but it looks like even setting `-Dorg.eclipse.swt.internal.carbon.smallFonts` doesn't bring that back. Your comment sounds like B) is the "new normal", but I highly doubt that users would prefer B) over A). Is there a way to bring back the look of A) for those elements somehow?

**Comment by Phillipus** [comment]
I'll try to explain the situation again (with reference to https://github.com/eclipse-platform/eclipse.platform.swt/issues/677)

- Since macOS 11 native row heights are taller (see [comment](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1846595468))
- Since Eclipse 4.30 both x86_64 and Silicon Mac Eclipse have a launcher built using macOS SDK 13 (see [comment](https://github.com/eclipse-platform/eclipse.platform.swt/pull/771#issuecomment-1889346638))
- When a new table/tree/list is created its row height is now taller (see [comment](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issue-1715165550))
- However when a table/tree/list has its font or image set it calls `setItemHeight` (see [here](https://github.com/eclipse-platform/eclipse.platform.swt/blob/fc6539786863fef89df6758dfaa51747e08a55fd/bundles/org.eclipse.swt/Eclipse%20SWT/cocoa/org/eclipse/swt/widgets/Tree.java#L3172) )
- Before the latest patch, `setItemHeight` added vertical padding of 1 pixel which is why you see smaller row heights in Eclipse 4.30. But actually the padding should be 8 pixels to be consistent with the new, taller, row heights which is why it's bigger in Eclipse 4.31

So the new row heights are actually correct for the new macOS SDK.

If `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set then `Tree#setFont()` is always called which in turn calls `setItemHeight` which sets the smaller item height in Eclipse 4.30 but not in 4.31.

- If you remove the `-Dorg.eclipse.swt.internal.carbon.smallFonts` argument the row heights you see in Eclipse are the new normal for Eclipse on Mac.

**Comment by Phillipus** [comment]
> I would actually prefer to continue to have the more compressed look of A) for all my views, lists, tress, etc., but it looks like even setting -Dorg.eclipse.swt.internal.carbon.smallFonts doesn't bring that back. Your comment sounds like B) is the "new normal", but I highly doubt that users would prefer B) over A). Is there a way to bring back the look of A) for those elements somehow?

Not without creating inconsistent behaviour. If the row height padding in `setItemHeight` was set back to 1 pixel then any RCP app, or Eclipse itself, that doesn't set `-Dorg.eclipse.swt.internal.carbon.smallFonts` (which initially forces a call to `setItemHeight`) would display the larger row height and the next time `setFont` was called on that control the item height would shrink (as explained [here](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issue-1715165550)).

To see this, add `System.out.println(getItemHeight());` as the first line in `Tree#setItemHeight()`. This will output `25`.

**Comment by Phillipus** [comment]
The only thing I can think of to maintain backward compatibility would be to use a vertical padding of 1 pixel if `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set and 8 otherwise.

**Comment by Phillipus** [comment]
Please comment and test https://github.com/eclipse-platform/eclipse.platform.swt/pull/1055

The caveat with this is that `-Dorg.eclipse.swt.internal.carbon.smallFonts` forces a call to `setItemHeight` which sets the row height smaller. The native initial row height is actually larger (25).

**Comment by martinlippert** [comment]
> The only thing I can think of to maintain backward compatibility would be to use a vertical padding of 1 pixel if `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set and 8 otherwise.

Yeah, I had the same thought, that would be awesome.

**Comment by Phillipus** [comment]
> > The only thing I can think of to maintain backward compatibility would be to use a vertical padding of 1 pixel if `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set and 8 otherwise.
> 
> Yeah, I had the same thought, that would be awesome.

@martinlippert Can you test https://github.com/eclipse-platform/eclipse.platform.swt/pull/1055?

**Comment by martinlippert** [comment]
> > > The only thing I can think of to maintain backward compatibility would be to use a vertical padding of 1 pixel if `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set and 8 otherwise.
> > 
> > 
> > Yeah, I had the same thought, that would be awesome.
> 
> @martinlippert Can you test [eclipse-platform/eclipse.platform.swt#1055](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1055)?

Yeah, happy to give this a try. Any recommendations about the easiest way to build this locally?

**Comment by Phillipus** [comment]
> Any recommendations about the easiest way to build this locally?

You need to clone https://github.com/eclipse-platform/eclipse.platform.swt and import the `org.eclipse.swt` and `binaries/org.eclipse.swt.cocoa.macosx.aarch64` projects into your workspace. You also need to fetch and checkout the PR.

**Comment by merks** [comment]
@martinlippert 

Using https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment fully automates the setup process.

It's a bit of overkill because it includes everything in the SDK, but it lets you launch a runtime workspace (launcher provide) where you can see the impact in context.

You can also go back:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/38d74be5-3651-4c2a-ad7c-84c4845a8bf9)

Deselect everything:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/cfba803f-b3af-4caf-a417-0342808265d5)

And select only SWT:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/1c574a65-fe17-4177-92b6-6a2577d757ec)

**Comment by Phillipus** [comment]
> The only thing I can think of to maintain backward compatibility would be to use a vertical padding of 1 pixel if -Dorg.eclipse.swt.internal.carbon.smallFonts is set and 8 otherwise.

Thinking about this more, this workaround is not really true to the intent of the `org.eclipse.swt.internal.carbon.smallFonts` setting.

With the workaround patch:

- If `smallFonts` is **not** set the default font is `.AppleSystemFont 13` and row height padding is 8 pixels.
- If `smallFonts` **is** set the default font is `.AppleSystemFont 11` and row height padding is 1 pixel
- However, in the case where `smallFonts` is **not** set and the user sets a control's font to `.AppleSystemFont 11` the row height padding is 8 pixels.

So with this patch we're effectively changing the meaning of the `smallFonts` setting to mean "use a smaller font **and** use a small row height". If we are to be true to the native OS, then really the larger row height should be used in all cases.

**Comment by merks** [comment]
I'm not on a Mac but if this issue were Windows I would be super annoyed to suddenly have a lower density of information content in all my views.

@martinlippert 

Are other places (other native mac applications) also showing such less dense tables, tree, and lists?

**Comment by Phillipus** [comment]
> I'm not on a Mac but if this issue were Windows I would be super annoyed to suddenly have a lower density of information content in all my views.

But this was actually the case for all native Mac apps when macOS Big Sur (11) was released. Since then native row heights on Mac are increased (if they target the later Mac SDK), as noted [here](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1846595468).

I've spent a **lot** of time thinking about, and experimenting with, this issue and have come to accept this as the new normal. In our RCP app, Archi, we don't set the `org.eclipse.swt.internal.carbon.smallFonts` setting. Thus, all row heights are now larger on later macOS. We can't change that, it's native, and SWT creates these controls with that increased row height. The problem for us was that controls had the larger row height but if a user changed the font for a control, it magically shrank the row height (because of the 1 pixel instead of 8 pixels in `setItemHeight`).

**Comment by Phillipus** [comment]
> Are other places (other native mac applications) also showing such less dense tables, tree, and lists?

Yes. Even the SWT-based app "SmartGit".

Before:

<img width="1035" alt="old" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/600504/7acee792-c49f-458f-9b23-1081832d0095">

After:

<img width="1042" alt="new" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/600504/10814cf1-bbf4-47aa-8b54-a5b6bbd068a9">

To be clear, if you don't use the `org.eclipse.swt.internal.carbon.smallFonts` setting and run Eclipse 4.30 on the latest macOS you'll see the increased row height (except for a few owner-draw tables and trees). This is the native row height. This [Snippet](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issue-1715165550) proves this.

**Comment by BeckerWdf** [comment]
> If we are to be true to the native OS, then really the larger row height should be used in all cases.

I can only second that. The operating system dictates how the high a line in a tree is. We should not try to circumvent this.
I think it's just a matter of what you are used to. When I first saw the differences I also thought that the old one was better. But the longer I see the new one I find the old one wrong - it's just too packed and too less whitespace.

**Comment by martinlippert** [comment]
> > Any recommendations about the easiest way to build this locally?
> 
> You need to clone https://github.com/eclipse-platform/eclipse.platform.swt and import the `org.eclipse.swt` and `binaries/org.eclipse.swt.cocoa.macosx.aarch64` projects into your workspace. You also need to fetch and checkout the PR.

Did all that, workspace looks okay, but starting up the runtime workbench gives me an error, complaining about the jnilib file

```
eclipse.platform.swt/binaries/org.eclipse.swt.cocoa.macosx.aarch64/libswt-cocoa-4964r8.jnilib' (not a mach-o file)
```

**Comment by Phillipus** [comment]
For me, it's a bit frustrating to explain all of this because the underlying cause is not simple. So I tend to repeat myself and continue to add more comments in the hope of clarifying the issue (even to myself). So...

- The look of Mac SWT controls is determined by what macOS SDK was used to build the launcher
- For Eclipse the launcher is the `eclipse` binary
- For pure SWT apps (and when launching a debug instance of Eclipse) the launcher is `java`
- Since Eclipse 4.30 the SDK used to build the launcher is SDK 13
- Therefore native row heights for the default font are 25

**Comment by Phillipus** [comment]
> Did all that, workspace looks okay, but starting up the runtime workbench gives me an error, complaining about the jnilib file
> 
> ```
> eclipse.platform.swt/binaries/org.eclipse.swt.cocoa.macosx.aarch64/libswt-cocoa-4964r8.jnilib' (not a mach-o file)
> ```

The SWT git repo uses git [LFS](https://git-lfs.com/) so you'll need to configure for LFS. This was changed [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/956).

**Comment by merks** [comment]
@martinlippert 

There is a reason why I suggested using the setup.  There are a bunch of "magical things" that need to be done that some folks know how to do, but it's poorly document, so mostly humans will make mistakes.

**Comment by martinlippert** [comment]
> I'm not on a Mac but if this issue were Windows I would be super annoyed to suddenly have a lower density of information content in all my views.
> 
> @martinlippert
> 
> Are other places (other native mac applications) also showing such less dense tables, tree, and lists?

Here is a screenshot from XCode, running on the same machine. You will notice that the font is slightly larger than in Eclipse (with the small font property)

<img width="268" alt="xcode-file-explorer-view" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/479b7f7c-78e0-45c1-908e-9083ff7f43ef">

Interesting outcome:
- if Eclipse uses a larger font (similar to what Xcode uses here), Eclipse has more space between rows
- if Eclipse uses a small font (with the option), it has about the same distance between rows than Xcode (using the larger font).

**Comment by Phillipus** [comment]
> > If we are to be true to the native OS, then really the larger row height should be used in all cases.
> 
> I can only second that. The operating system dictates how the high a line in a tree is. We should not try to circumvent this. I think it's just a matter of what you are used to. When I first saw the differences I also thought that the old one was better. But the longer I see the new one I find the old one wrong - it's just too packed and too less whitespace.

I agree. My vote is to **not** merge the draft PR I created.

**Comment by BeckerWdf** [comment]
> You will notice that the font is slightly larger

I think that this is true but to be honest I cannot *see* that. To see that really I would like to see screenshots next to each other.

**Comment by Phillipus** [comment]
Checking the row heights on native Mac apps is not a good indicator. We don't know what an app is doing and whether they have custom drawing for controls, or if it targets a newer or old Mac SDK. In fact, even in Eclipse itself, you'll see some controls in Preferences that have smaller row heights because they do some custom drawing. What we can say with certainty is that when SWT creates a new tree/table/list the row height is natively larger (the `smallFonts` setting clobbered that in `setItemHeight`)

**Comment by Phillipus** [comment]
I'm running out of steam with this issue. ;-). Can we have a consensus?

- Because of fixing a legacy padding value [in this PR](https://github.com/eclipse-platform/eclipse.platform.swt/pull/771) row heights will be larger with the `smallFonts` setting. The reason why we saw smaller row heights when `smallFonts` was set was a side-effect of a control calling `setItemHeight` with a padding of 1 pixel immediately after their creation.
- The native and correct row height is visible when the `smallFonts` is **not** set
- To accurately set the row height in all cases the padding should be 8 pixels

Thoughts?

**Comment by martinlippert** [comment]
> > You will notice that the font is slightly larger
> 
> I think that this is true but to be honest I cannot _see_ that. To see that really I would like to see screenshots next to each other.

Since you asked for it, here we are. I shot a few side-by-side screenshots. Each one is comparing 2024-03 CI with small fonts on the left, without small fonts in the middle, and with some other macOS app on the right.

Compared to Eclipse 2023-12 (where 2023-12 has smallfonts enabled):
<img width="908" alt="side-by-side-eclipse" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/2066c3b2-8260-44d2-a750-7d6f2dfe2f06">

Compared to VSCode (default settings in VSCode, no zoom):
<img width="966" alt="side-by-side-vscode" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/74892392-e4d9-456b-8f8e-aa6b40ab8586">

Compared to Xcode:
<img width="881" alt="side-by-side-xcode" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/e7873993-2190-4ea2-9681-b7cbba411220">

Compared to macOS native finder app:
<img width="959" alt="side-by-side-finder" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/107705/3aac4ff1-83b5-4760-bf84-c5497de62623">

**Comment by martinlippert** [comment]
I don't know a lot about the implementation details and I am grateful for all the technical details provided here and the discussion around that, very very much appreciated. I am just looking at this from the point of view of the developers using Eclipse as their IDE on macOS and this change that is coming to them, having the quote from Ed in mind:

> I'm not on a Mac but if this issue were Windows I would be super annoyed to suddenly have a lower density of information content in all my views.

**Comment by Phillipus** [comment]
As I [said earlier](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-1954249566), comparing with XCode or Finder is meaningless.

Please take another look at the first comment and snippet [here](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677). When a SWT Tree/Table/List is created its row height is **25 pixels** (assuming the standard default macOS font). That is correct for macOS SDK creating a new control for SWT. Now in order to maintain the correct padding in a row when `setItemHeight` is called (triggered by `setFont`) the padding has to be **8 pixels**, not 1 pixel as it was before.

I don't know how I can say it any other way - the old value of 1 pixel in `setItemHeight` is **wrong** for later macOS versions.

**Comment by Phillipus** [comment]
> I don't know a lot about the implementation details and I am grateful for all the technical details provided here and the discussion around that, very very much appreciated. I am just looking at this from the point of view of the developers using Eclipse as their IDE on macOS and this change that is coming to them, having the quote from Ed in mind:
> 
> > I'm not on a Mac but if this issue were Windows I would be super annoyed to suddenly have a lower density of information content in all my views.

Then the blame has to go to Apple. ;-)

**Comment by martinlippert** [comment]
> > > The only thing I can think of to maintain backward compatibility would be to use a vertical padding of 1 pixel if `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set and 8 otherwise.
> > 
> > 
> > Yeah, I had the same thought, that would be awesome.
> 
> @martinlippert Can you test [eclipse-platform/eclipse.platform.swt#1055](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1055)?

Thanks a lot again for helping me setting up an environment to test this, I finally succeeded... :-) Many many thanks again to you all, much appreciated.

And yes, I can confirm that this brings back the look of Eclipse 2023-12 when running with the smallfonts setting enabled (as all the Eclipse IDE packages do, as far as I am aware of).

I know there are different opinions about what the right approach here is and I do understand the points being made. My personal vote would be to merge this PR in [eclipse-platform/eclipse.platform.swt#1055](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1055.

**Comment by martinlippert** [comment]
(and I promise, I will not post any screenshots anymore) 😂

**Comment by Phillipus** [comment]
I recognise that the increased row height might be unwelcome for some Mac users. But consider those users who don't use the `smallFonts` setting either in Eclipse or an SWT based app. Our RCP app has never used that setting and so our users are going to get a surprise in our next version when row heights are bigger. There's nothing I can do about that.

**Comment by merks** [comment]
It's rather late in the development cycle to suggest yet another system property, and if we don't do that this release cycle, everyone will just get used to it such that changing it again would be kind of strange.

I have no strong opinion of what is right or best for the general public.  There are good arguments made in favor of doing what is natively intended...

**Comment by Phillipus** [comment]
> to suggest yet another system property

I already thought of, and tested, that idea.

In the `Display` class add:

```java
boolean smallItems;

// Initialise it somewhere
smallItems = System.getProperty("org.eclipse.swt.smallItems") != null;
```

In Table:

```java

// Have to initialise item height in constructor
public Table (Composite parent, int style) {
    super (parent, checkStyle (style));
    setItemHeight(null, null, true);
}

// Check display.smallItems here...
void setItemHeight (Image image, NSFont font, boolean set) {
    if (font == null) font = getFont ().handle;
    double ascent = font.ascender ();
    double descent = -font.descender () + font.leading ();
    int height = (int)Math.ceil (ascent + descent) + (display.smallItems ? 1 : VERTICAL_CELL_PADDING);
...

```

**Comment by Phillipus** [comment]
Are there other Mac devs we can ping to get an opinion? @HannesWell @HeikoKlare do you have Mac access?

**Comment by akurtakov** [comment]
@lshanmug wdyt?

**Comment by HeikoKlare** [comment]
Even though I have Mac access, I am not a "real Mac user", as I only use it for Eclipse development validation purposes. So probably my opinion on this is not that relevant and I actually do not have a strong opinion on the specific issue here.

I see a more general issue here: there is no specification and no common agreement on what is the "baseline" for how the Eclipse/SWT UI should look like on each OS. I understand every point of view or preference, be it that someone is used to a specific look & feel, be it what the underlying framework/OS etc. gives you, or be it whatever makes "most sense" for some other reasons (like a cleaner look with larger spaces or more information with smaller spaces). I don't see that one can be objectively better than the other, but it will just be about personal preference. Mine is as follows: I do not really see anything feeling "native" anymore (which once was the baseline for SWT look & feel). In particular on Windows, Eclipse, even though using native graphics, feels like the only application looking that way (as the native API is 20+ years old). I am not sure how the feeling is for Linux or Mac users, but there are probably more applications using GTK and Cocoa looking kind of similiar. I would definitely prefer if we had an "Eclipse look & feel" rather than a "native look & feel". But that would require users of some OS leave behind what they are used to, since the look & feel is quite different between the OS now, in particular regarding spaces that are small on Windows, large on GTK and obviously changing on Mac.

Sorry for only being able to make such a general comment. Trying to apply it this issue I would say that sticking to existing behavior (and maybe making it more configurable) is more valuable than adopting something "native".

**Comment by HannesWell** [comment]
> Are there other Mac devs we can ping to get an opinion? @HannesWell @HeikoKlare do you have Mac access?

Sorry I don't have Mac access, I'm on Windows at home and on Linux at work.

**Comment by Phillipus** [comment]
@HeikoKlare Thanks for your feedback.

The underlying issue here is not necessarily to change things but to be _consistent_. Again, I refer to this [comment](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issue-1715165550) and the [Snippet](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1552648371) there. When a Table/Tree/List is created in SWT it has the larger row height (because it is created natively by the OS). Previously, calling `setFont()` (which in turn calls `setItemHeight()`) on the control would reduce the row height because of using 1 pixel as vertical padding. Now, to be consistent with the larger initial row height, we use a value of 8 pixels.

Even before the change to 8 pixels in `setItemHeight()` if you create an RCP application that does not call `setFont` on the tree/table/list you get the bigger row height whether you like it or not. What you don't want is the row height magically shrinking when and if you do call `setFont` (as is the case if `-Dorg.eclipse.swt.internal.carbon.smallFonts` is set).

What we _can't_ do is to go back to using 1 pixel as the vertical padding for the reasons just stated. Either ensure these controls are created with a smaller initial height (I don't know how you'd do that as the control is created natively) or leave this as it is.

And see https://wiki.lazarus.freepascal.org/macOS_Big_Sur_changes_for_developers#Table_Views

**Comment by lshanmug** [comment]
Thanks @Phillipus for the detailed description about the new behavior. 

The increased row height for `Table/Tree` is the new default behaviour that was introduced with macOS BigSur (in Nov 2020). So, SWT applications/snippets on aarch64 were already seeing increased row height since then. For example, https://bugs.eclipse.org/bugs/show_bug.cgi?id=575696. Eclipse by default was kind of shielded from this change so far because it had `-Dorg.eclipse.swt.internal.carbon.smallFonts` specified in the eclipse.ini which controls the font size. However, even with older Eclipse builds, launching the child eclipse had the taller rows as the launch config doesn't have `-Dorg.eclipse.swt.internal.carbon.smallFonts` specified.

Looks like with the fix for https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#,  `-Dorg.eclipse.swt.internal.carbon.smallFonts` has lost it's 'special effect' on the row height. I've not looked into the code in detail to see how it can be done, but if we can retain the effect of the smallFonts property things could be same as before from both Eclipse and SWT POV.

@Phillipus Please correct me if I'm missing something in my comment.

**Comment by soethoff** [comment]
macOS dev here: I always used the `-Dorg.eclipse.swt.internal.carbon.smallFonts` parameter (I think it's set per default, at least for the Java Developers package), so even on aarch64 I didn't see any change, until now.

> It's rather late in the development cycle to suggest yet another system property, and if we don't do that this release cycle, everyone will just get used to it such that changing it again would be kind of strange.

Agree, first I was confused, now I got used to and would prefer to keep it 'native'. It's now behaving the same as other IDEs like VSCode or IntelliJ (excluding Xcode or Finder), at least when still using the `-Dorg.eclipse.swt.internal.carbon.smallFonts` parameter. I also have the feeling, that it looks now more 'natural', but sure, this is very personal.

![eclipse_vscode](https://github.com/eclipse-platform/eclipse.platform.ui/assets/34278657/1c92a7a3-0edd-44f5-bf50-9add572ab2e3)

If there is the wish to have something like this configurable, it might make sense to offer some new appearance option like macOS does (system settings -> more space / bigger letters). In that way the overall look & feel adjusts to the wished / needed scaling and not only the font size nor the padding.

> I agree. My vote is to not merge the draft PR I created.

I also like to vote to not merge the draft PR.
My understanding would be that with this change it would not be possible anymore to have the 'native' padding and small fonts at the same time.

**Comment by Phillipus** [comment]
Thanks for the feedback @lshanmug your analysis is correct. In fact the reason why we only saw increased row height in aarch64 Mac version and not x86_64 version was not because of the architecture difference but because the aarch64 launcher was built with Mac SDK 11 and the x86_64 version launcher was built with Mac SDK 10. Now that both launchers are built with SDK 13 the row height increase is present in both versions.

Therefore, the increased vertical padding of 8 pixels in the `setItemHeight` method is absolutely correct in all cases to maintain consistency and to stop row heights magically shrinking when the font is set. See my comment [here](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-1954784827).

This idea from @soethoff seem more in tune with the intent:

> If there is the wish to have something like this configurable, it might make sense to offer some new appearance option like macOS does (system settings -> more space / bigger letters). In that way the overall look & feel adjusts to the wished / needed scaling and not only the font size nor the padding.

`org.eclipse.swt.internal.carbon.smallFonts` should mean just that, use a smaller font (I think its font height 11 instead of 13), not remove additional native padding.

> I also like to vote to not merge the draft PR.
My understanding would be that with this change it would not be possible anymore to have the 'native' padding and small fonts at the same time.

Exactly.

When Apple released Big Sur 11 they defined [new UI metrics](https://wiki.lazarus.freepascal.org/macOS_Big_Sur_changes_for_developers#Table_Views).

**Comment by Phillipus** [comment]
POC idea to add (yet another ;-) ) startup option for smaller item (row) heights. This one is separate to the `smallfonts` option and so it would be possible to have the 'native' padding and small fonts at the same time.

https://github.com/eclipse-platform/eclipse.platform.swt/pull/1057

**Comment by a1dutch** [comment]
I've just updated to Eclipse 2024-03 and the extra whitespace is not nice in the package explorer especially

**Comment by Phillipus** [comment]
> I've just updated to Eclipse 2024-03 and the extra whitespace is not nice in the package explorer especially

The reason why item heights are larger has been made clear several times in this thread. The underlying cause is due to Apple's changes to native UI elements since Big Sur. A bug fix in SWT now corrects the true item height when the font is set so that they are in line with the true native height.

What we can't have is a Tree/Table/List starting out with the increased item height when it is created and then magically shrinking when the font on that control is set. Eclipse IDE users don't see the initial larger row height because of the `org.eclipse.swt.internal.carbon.smallFonts` setting forcing the smaller height after creation when `setFont` is called (a bug), but you will see the larger heights in an RCP app or debugged child Eclipse when it is not set.

**Comment by kohlschuetter** [comment]
I just updated to 2024-03 on macOS and also very much noticed the increased line height.

I have a lot of projects open at the same time, and I rely on the high information density as it was with Eclipse 2023-12 and before.

Adding this line space reduces the chance that I immediately see certain project warnings (yellow icons in the Project Explorer). I also have to scroll more, which makes me very sad.

The argument I read so far is "it works as expected because Apple decided so in an SDK change, and we now happen to use the new SDK".

I totally understand the rationale for following UI guidelines for regular UI elements such as settings.

However, as a user, I would like to be able to customize the appearance to my needs, and it looks like whatever reason the change in 2024-03 is, to me it is a regression.

Notably, even Apple themselves don't follow these rules themselves blindly. Take Xcode as an example.
They support three levels of spacing "Normal" but also "Tight" and "Relaxed", at least for the text editor.

For the project/source code explorer view, Xcode also seems to be able to show 13pt fonts in the same height as Eclipse does for 11pt fonts. (If you increase Eclipse's font size in "Tree and Table font for views" to 13pt, the height is a few pixels larger than Xcode for the same font size).

**Comment by Phillipus** [comment]
@kohlschuetter You need to understand the real reason (made clear in this thread several times) about why this change was made.

When SWT creates a new Table, Tree, or List _it is created natively with the increased row height_ (because it is using macOS SDK 11 or greater). It is a side effect and bug that when the option `org.eclipse.swt.internal.carbon.smallFonts` is used it forces a call to `setItemHeight` which used to add an (incorrect) padding of 1 pixel, hence the smaller row heights. To maintain consistency with the initial row height the correct padding is now set to 8 pixels. Without this change, an RCP app or SWT-based app that creates one of these components will initially display with the larger row height and will then shrink later whenever `setItemHeight` is called.

Not all apps (and launching child Eclipse) use the `org.eclipse.swt.internal.carbon.smallFonts` setting and so even before this change there were increased row heights (in Eclipse 4.30 remove the `org.eclipse.swt.internal.carbon.smallFonts` setting from the eclipse.ini file and see the row heights are bigger)

The issue is to maintain consistency between the initial row height when a control is created, and when `setItemHeight` is called.

I have proposed a new option for smaller row heights [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1057).

**Comment by merks** [comment]
As we see, however correct it may be, users perceive it as different and not better.  😞

**Comment by exabrial** [comment]
> As we see, however correct it may be, users perceive it as different and not better. 😞

UI issues are frustrating in that regard, as it's difficult to separate preference vs performance. From my perspective, being able to see roughly 1/3 the information I used to be able to, and instead stare at blank space, is a regression. Being reasonable and after reading through this thread, I'm willing to admit that this could be "preference" however.

It would be a great relief to make this a configuration item for those who prefer more information on their screen.

As always, thank you Eclipse Team, happy Eclipse user here. I'm happy to help any way I can.

**Comment by merks** [comment]
@exabrial 

That a super nice way to express it!  If only more people were so reasonable, understanding, and nice.  ❤️

**Comment by Phillipus** [comment]
> As we see, however correct it may be, users perceive it as different and not better. 😞

And I agree with that. But we have to find a solution that fixes the discrepancy in initial row heights as created by SWT and the "magical shrinkage" that can occur later, and also provide a means to restore the older item heights.

The padding used in [`setitemHeight`](https://github.com/eclipse-platform/eclipse.platform.swt/blob/9149b68f6d735c11f0eb0ca45dc8717c2e265760/bundles/org.eclipse.swt/Eclipse%20SWT/cocoa/org/eclipse/swt/widgets/Tree.java#L3176) has to match that of the control's row height when it is created. Otherwise we get "magical shrinking" of row heights. Either these controls are initially created with a smaller row height or a bigger row height and that row height is maintained when `setitemHeight` is called. Like it or not, the old way led to [this bug](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677).

What I'm saying is this:

**If Mac users want the smaller row height then it should be set when a Tree/Table/List is initially created, not "magically shrunk" by chance if and when `setitemHeight` is called**. This [PR](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1057) does that.

So, we need to address two concerns:

1. Maintain consistency of row height when a control is created and when `setitemHeight` is called
2. Either use small row heights or native macOS row heights, or add an option to use either of these

**Comment by kohlschuetter** [comment]
@Phillipus Thanks for providing the patch! I tried it and also experimented with different values for that vertical cell padding.

The smallest value that made a difference for my 11 point default font was 3 (not 1), bringing back what looks like the original line spacing. 5 looked actually not too bad, and the default 8 is definitely too large to me.

Concerning the bug in #677 (setting the same font again reduces item height), do I understand correctly that this is mitigated by `setFont((Font)null);` in your patch? If not, would setting `VERTICAL_CELL_PADDING` to `3` fix it?

Moreover, we could actually use a factor for that "vertical cell padding" instead of an absolute value, because it's actually just a line "leading" property (which should be around 20-30% of the font size. That way, it will work correctly for larger font sizes, too (a factor of 1.2 is like the old behavior; 1.35 looks still OK for all sizes and appears to result in the same spacing as the project explorer in Xcode with font size 13pt)

**Comment by Phillipus** [comment]
> Concerning the bug in https://github.com/eclipse-platform/eclipse.platform.swt/issues/677 (setting the same font again reduces item height), do I understand correctly that this is mitigated by setFont((Font)null); in your patch? If not, would setting VERTICAL_CELL_PADDING to 3 fix it?

Calling `setFont((Font)null);` in `List` and `setItemHeight` in `Table` and `Tree` sets the item heights in the constructor so that they are consistent.

Please note, as I keep trying to emphasise, the problem is fixing the discrepancy between the initial item height when a control is created and the item height calculated in `setitemHeight` (or `setFont` in the `List` control)

> we could actually use a factor for that "vertical cell padding" instead of an absolute value,

I tried that and it led to differing results between the native initial item height and the one calculated in `setitemHeight`. Perhaps it needs some fine tuning.

**Comment by Phillipus** [comment]
I really don't know how I can make the issue clearer. I must have have explained it here and in other issues perhaps 20 or so times? What part do people not understand? (Genuine question, not sarcasm).

**Comment by kohlschuetter** [comment]
@Phillipus I think the mismatch in our communication is
1. what users like me describe as a problem here (call it the "too much space problem") and
2. what you refer to as the problem (the problem that was addressed with the patch for #677, which caused the visual regression — call it the "height gets smaller problem").

These appear to be two different problems. I haven't seen problem 2 myself, but I'm not sure I would agree that hard-coding an additional 7 pixels for padding sounds like a proper solution either. It feels like there's more to that to understand what's going on.

I looked into your demo code in #677 and tried to reproduce the bug on my machine (macOS Sonoma 14.4 (23E214) MacBook Pro M1 Max).

To do so, I created a new Java project, linking org.eclipse.jface_3.33.0.v20240214-1640.jar from Eclipse 2024-03 and then either Eclipse's org.eclipse.swt.cocoa.macosx.aarch64 jar or the SNAPSHOT one from git (using  eclipse.platform.swt branch `R4_31`, applying your POC commit on top of that), and using a `-Djava.library.path=/Applications/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/2599/0/.cp` for the native dylib. For the SNAPSHOT builds I tested both with and without `-Dorg.eclipse.swt.internal.cocoa.smallItemHeight` 

I clicked on the "Set font" button, but in all three cases the table item height did _not_ change.

The only difference in the three runs being the height itself (18 with snapshot and `-Dorg.eclipse.swt.internal.cocoa.smallItemHeight`, 25 otherwise).

Adding `-Dorg.eclipse.swt.internal.carbon.smallFonts` also didn't trigger the bug for me, resulting a font height of 22 (15 with `smallItemHeight`)

Could it be that you are seeing a problem that was introduced and then subsequently fixed in Cocoa/AppKit? 

If so, why bother? The cure is much worse than the disease, so to speak.

**Comment by Phillipus** [comment]
> I clicked on the "Set font" button, but in all three cases the table item height did not change.

(The link to that issue is actually https://github.com/eclipse-platform/eclipse.platform.swt/issues/677)

Yes, because that was fixed [here](https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d)  and so you won't see the problem unless you use an earlier version of SWT. And that fix leads us here...

> Could it be that you are seeing a problem that was introduced and then subsequently fixed in Cocoa/AppKit?

No. Please read all of my analyses here.

**Comment by Phillipus** [comment]
Either revert https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d (and go back to inconsistent and broken behaviour) or follow through with a solution that attempts to maintain the current consistent behaviour _and_ also provides an option for smaller item heights as well as native item heights as proposed [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1057). i.e. solves both issues.

**Comment by kohlschuetter** [comment]
OK, I reverted [eclipse-platform/eclipse.platform.swt@0032340](https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d), and I can reproduce the "height gets smaller problem".

Then I added `setItemHeight(null, null, true);` to the `Table` constructor (like what you did in the POC code, sans the configurable padding logic), i.e.

```
public Table (Composite parent, int style) {
	super (parent, checkStyle (style));

    setItemHeight(null, null, true);
}
```

and the item height reliably stays at 18 when I click "Set font"

The list looks OK visually, isn't that all we need?

<img width="612" alt="Screenshot 2024-03-15 at 00 18 16" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/822690/153634ce-20fe-4419-a662-a1478ce2c597">

I understand you also want to be able to support whatever "native" item heights are, but obviously that number depends on some SDK version and is not relevant to any previously working Eclipse app.

Maybe that "feature" can be indeed configured separately? For example, you could simply retrieve the native height in the constructor, via `((NSTableView)view).rowHeight()` (which returns 24 instead of 17 on my M1 Mac) and use that as a "minimal height" constraint that can be enable/disabled independently...

Now back to the "Project Explorer looks bad" problem: Yes, you could make that "set minimal height to system default" the default case for Eclipse IDE (maybe via a boolean System property `enforceNativeHeights=true` or something like that), adding an exception for the Project Explorer (and all other places that matter), allowing whatever value `setItemHeight` determines today. That way, all other system UI in Eclipse, like the Preferences etc., will retain the "native macOS look". Is that more like what you want?

**Comment by Phillipus** [comment]
> The list looks OK visually, isn't that all we need?

No.

**The native (and correct) item height for the control is the larger one when it is first created.** When `setItemHeight` is called it made it smaller in pervious versions because of the 1 pixel padding. In 4.31 a correct 8 pixels of padding is used. This is the correct behaviour. The aim is to maintain the same item height when the control is created and when `setItemHeight` is called.

If people want the smaller item height then it has to be set both in the control's constructor and in `setItemHeight`. This should be optional so that Eclipse and RCP apps can use the native item height.

- Get a copy of Eclipse 4.30 for Mac (i.e before [this change](https://github.com/eclipse-platform/eclipse.platform.swt/commit/0032340b9b647db3ef6776d6b2e481397bd74b8d) )
- Launch Eclipse 4.30 to see small item heights
- Edit the `Eclipse.app/Contents/Eclipse/eclipse.ini` file and remove the `-Dorg.eclipse.swt.internal.carbon.smallFonts` setting
- Launch Eclipse again to see larger item heights

The smaller heights are a side effect of the `-Dorg.eclipse.swt.internal.carbon.smallFonts` setting calling `setItemHeight`. Many RCP apps don't use this setting and so get the native, bigger, row heights.

**Comment by Phillipus** [comment]
> Now back to the "Project Explorer looks bad" problem: Yes, you could make that "set minimal height to system default" the default case, but then add an exception for the Project Explorer, allowing whatever value setItemHeight determines today. That way, all other system UI in Eclipse, like the Preferences etc., will retain the "native macOS look". Is that more like what you want?

No.

One more time:

The native (and correct) item height for the control is the larger one when it is first created. When setItemHeight is called it made it smaller in pervious versions because of the 1 pixel padding. In 4.31 a correct 8 pixels of padding is used. This is the correct behaviour. The aim is to maintain the same item height when the control is created and when setItemHeight is called.

If people want the smaller item height then it has to be set both in the control's constructor and in setItemHeight. This should be optional so that Eclipse and RCP apps can use the native item height.

**Comment by Phillipus** [comment]
I can't keep on repeating myself and banging my head against a brick wall.

**Just make item heights consistently the same when they are created and when the item height is calculated in `setItemHeight`.**

That's it.

I don't care what the item height is as long as it's the same in both places.

**Comment by exabrial** [comment]
Phillipus thank you for taking the time to help everyone work towards the most correct solution. Sorry it's frustrating!

**Comment by Phillipus** [comment]
> Phillipus thank you for taking the time to help everyone work towards the most correct solution. Sorry it's frustrating!

Thanks.

I'm going to make yet another attempt to explain the situation...

1. Some time ago, Eclipse for Mac was compiled/linked to macOS 10 and so the native row heights were small when these controls were instantiated.
2. Now that Eclipse for Mac is compiled/linked to macOS 13 the native row heights are larger to match the UI of the linked macOS SDK.
3. After (2) Eclipse users continued to see small row heights because of the (incorrect) 1 pixel padding applied when `setItemHeight` was called (triggered by `-Dorg.eclipse.swt.internal.carbon.smallFonts` and `setFont`).
4. However, this 1 pixel padding didn't match the initial native row height when the controls where instantiated and was changed to 8 pixels in a recent fix that made it into Eclipse 4.31 (see [here](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677))
6. After the fix in (4) Eclipse users now see the larger row heights because of this padding correction. (They would have seen these larger row heights before if `-Dorg.eclipse.swt.internal.carbon.smallFonts` was not applied)
7. Users don't like the (technically correct) increased row heights

So, the way forward is to:

1. Maintain the same row height height when a control is instantiated _and_ when `setItemHeight` is called
2. Provide a way for the row heights to be small (old behaviour) _and_ native sizes (macOS UI behaviour)

**Comment by kohlschuetter** [comment]
@Phillipus Please take a look at my pull request. I hope I've covered all cases.
Notably, the change does not make any assumptions about the vertical padding being 8, it just enforces the default height as the minimum if required. This also means we don't have to tune any parameters, and even retain compatibility with older, future, and unexpected macOS settings.

I verified the patch with your TableFont test, and I consistently get an item height of 25 with no configuration on newer macOS (and likely 18 on older macOS), an item height of 15 with (`-Dorg.eclipse.swt.internal.carbon.smallFonts`, as Eclipse IDE has by default, bringing back the expected behavior from 2023-12 and before), or a height of 18 for those folks who want the old macOS behavior back but don't want smallFonts.

Did I miss any scenario?

https://github.com/eclipse-platform/eclipse.platform.swt/pull/1117

**Comment by Phillipus** [comment]
> @Phillipus Please take a look at my pull request. I hope I've covered all cases.

https://github.com/eclipse-platform/eclipse.platform.swt/pull/1117

This is now the PR that is being worked on to come up with a solution, thanks to @kohlschuetter

**Comment by kohlschuetter** [comment]
Until the fix from https://github.com/eclipse-platform/eclipse.platform.swt/pull/1117 is merged, feel free to try this binary as a hot fix (copy to `/Applications/Eclipse.app/Contents/Eclipse/plugins/org.eclipse.swt.cocoa.macosx.aarch64_3.125.0.v20240227-1638.jar` or wherever you placed your Eclipse app).

The jar in hotfix.zip should replace the existing jar with the exact same name.

[hotfix.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/14623693/hotfix.zip)

**Comment by exabrial** [comment]
Just wanted to report back no issues with the hotfix initially!  @Phillipus thank you for guiding the discussion onto the right track and @kohlschuetter for providing the code. Very much appreciated.

**Comment by MahatmaFatalError** [comment]
So, 76 comments later, what is the conclusion for the just released 2024-03 ?

**Comment by martinlippert** [comment]
I think the fix will be part of the `2024-06` release. If you want to fix this in a running `2024-03` installation, I would follow the steps described in https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-2001986963. Worked for me quite nicely.

**Comment by ptrussart** [comment]
Is there a way to have a fix for the x86_64 Mac processor ?

**Comment by tjmackert** [comment]
> Until the fix from [eclipse-platform/eclipse.platform.swt#1117](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1117) is merged, feel free to try this binary as a hot fix (copy to `/Applications/Eclipse.app/Contents/Eclipse/plugins/org.eclipse.swt.cocoa.macosx.aarch64_3.125.0.v20240227-1638.jar` or wherever you placed your Eclipse app).
> 
> The jar in hotfix.zip should replace the existing jar with the exact same name.
> 
> [hotfix.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/14623693/hotfix.zip)

Is there an X86_64 hotfix available as well?

**Comment by laeubi** [comment]
I just wanted to note that putting arbitrary "hotfix jars" downloaded from the internet  into your eclipse folder can impose server security risk to your eclipse installation, therefore I would recommend to either build it locally yourself or use the official I-builds:

https://download.eclipse.org/eclipse/downloads/

The integration builds will also contain SWT jars you can replace then in your installation.

**Comment by Sciuriware** [comment]
Like to add: "Haste makes waste"
;JOOP!

**Comment by kohlschuetter** [comment]
I personally would not recommend anyone to upgrade their stable Eclipse release to an integration build with probably hundreds of unrelated changes.

@merks @lshanmug Can we please get an official release for the fix, without having to update to an unstable integration build?

Furthermore, I get the whole xz/liblzma story may cause some folks to go extra paranoid but I want to set things straight here:

I by myself built and provided the jar built from the code I wrote myself in the PR. The PR and jar is hosted on github.com. I use my real name in these communications, and with that I hopefully still have a reputation to lose. 

@laeubi:

I feel that you calling this work an "arbitrary hotfix jar downloaded from the internet" that "can impose server security risk to your eclipse installation", is an insult to outside contributors like me, and a disservice to all open source contributors that act in good faith. I have attached the jar to this patch solely to help people in the moment, and I think you know that.

On that note: I may also remind you that before this fix I have also fixed https://github.com/eclipse-m2e/m2e-core/issues/1511 which was broken by a change that you introduced yourself almost a year before, and it was probably not fixed earlier because you (erroneously) claimed that another upstream bug was to blame. Since I see a pattern here, maybe you should lower your accusing tone a bit?

I have spent numerous hours chasing both bugs and patiently followed up getting these fixes in so others can also benefit from my findings. Please, without diverting into the general case, have at least the decency to respect my contribution in good faith, or show us objective proof that my hotfix jar is a security risk.

As to this date, updating Eclipse from within its own app occasionally shows me unsigned plugins that either have to trust blindly or not install at all, because the existing landscape of Eclipse plugins, integration build update sites and the Eclipse marketplace are a shit show by themselves.

I mean, even for the signed plugins, I can't really tell the quality of a release, especially when obviously visible bugs like this present one don't qualify for a rather immediate patch release.

This is something you should look into before starting the scaremongering.

**Comment by merks** [comment]
The next official release is in June.  So unfortunately folks need to wait until that release or consume integration/milestone builds to get it sooner.  The release engineering effort to produce a release of the platform is already daunting. Backporting the fix to the 4.31 maintenance branch is feasible but you will notice that there are no published builds for that stream   Providing a fix for just this issue for the IDE packages on the download server just doesn’t seem feasible logistically. 

You certainly have my gratitude and respect for helping to resolve this bug. It’s effort beyond the call of duty.  If even 1 in 10,000 were this helpful, we’d be flooded in riches. ❤️

**Comment by kohlschuetter** [comment]
Thanks a lot @merks, I appreciate your kind words.

**Comment by laeubi** [comment]
> I personally would not recommend anyone to upgrade their stable Eclipse release to an integration build with probably hundreds of unrelated changes.

No one needs to upgrade their whole "stable Eclipse Release", Eclipse installations can perfectly run side-by-side. Also this integration builds will become the next release so if people would _test_ these beforehand one probably would identify the one of the hundreds maybe surprising or even harmful changes _before_ the release what would prevent such "after the fact"  discovery that is hard to fix as @merks has mentioned.

Beside that I never claimed _your_ patch / jar is harmful, nor did I insulted / addressed _you_ in anyway. 
It was just a general warning because in this thread people seem to tend to recommend this practice and there was even a follow up request for another patch-jar. And what if now I would upload such a patch  x86_64 Mac processor to github then people will probably using it because it was done so in the past. **If that ever happens please don't do that!** My computer might be compromised, my github account might be hacked, someone created an account that looks similar to mine, or whatever else.

This simply bypasses even the most basic checks we have in Eclipse/P2 and has nothing to do with being paranoid or scaremongering. Also this does not mean your contribution is bad in anyway or not appreciated.

**Comment by martinlippert** [comment]
> Also this integration builds will become the next release so if people would _test_ these beforehand one probably would identify the one of the hundreds maybe surprising or even harmful changes _before_ the release what would prevent such "after the fact" discovery that is hard to fix as @merks has mentioned.

I fully agree to this, but I also feel the need to mention that I discovered and reported this specific issue _before_ the release (using M2). So this was not a discovery _after the fact_  ;-)

**Comment by laeubi** [comment]
> > Also this integration builds will become the next release so if people would _test_ these beforehand one probably would identify the one of the hundreds maybe surprising or even harmful changes _before_ the release what would prevent such "after the fact" discovery that is hard to fix as @merks has mentioned.
> 
> I fully agree to this, but I also feel the need to mention that I discovered and reported this specific issue _before_ the release (using M2). So this was not a discovery _after the fact_ ;-)

Yep and it was worked on so everything is fine for me, it was also more a general comment and directed to the "probably hundreds of unrelated changes", the more people test it the earlier they we detect problem, the more likely one gets a fix / better quality in release. 

So from my experience ibuilds are very stable or if there is a problem are fixed sometimes even withing hours, so if I'm affected by a bug introduced in a release I can either stay at the old or use the just fixed ibuild and see if something else will affect my workflow.

**Comment by martinlippert** [comment]
> Until the fix from [eclipse-platform/eclipse.platform.swt#1117](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1117) is merged, feel free to try this binary as a hot fix (copy to `/Applications/Eclipse.app/Contents/Eclipse/plugins/org.eclipse.swt.cocoa.macosx.aarch64_3.125.0.v20240227-1638.jar` or wherever you placed your Eclipse app).
> 
> The jar in hotfix.zip should replace the existing jar with the exact same name.
> 
> [hotfix.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/14623693/hotfix.zip)

Hey @kohlschuetter, can you explain to me how you've exactly built this hotfix? I would like to run that myself, so in case you have the command that you used at hand or can share the script (or whatever you used), that would be awesome. Many many many thanks upfront!

**Comment by kohlschuetter** [comment]
@martinlippert If I remember correctly, it's just a matter of running `mvn clean install -DskipTests` on my branch of `eclipse.platform.swt.git` and then `cp binaries/org.eclipse.swt.cocoa.macosx.aarch64/target/org.eclipse.swt.cocoa.macosx.aarch64-3.126.0-SNAPSHOT.jar /Applications/Eclipse.app/Contents/Eclipse/plugins/org.eclipse.swt.cocoa.macosx.aarch64_3.125.0.v20240227-1638.jar`  (replace `aarch64` with `x86_64` if necessary)

Please work with code obtained via `git checkout 611accb844fb525efb3727f7fe8e2e7a3472debf` from https://github.com/eclipse-platform/eclipse.platform.swt.git. Note that the PR has evolved since the hotfix version (properties have been renamed etc. ) so you will not get exactly the same binary.

**Comment by HannesWell** [comment]
> I personally would not recommend anyone to upgrade their stable Eclipse release to an integration build with probably hundreds of unrelated changes.

In case you are concerned that for example https://download.eclipse.org/eclipse/updates/4.32-I-builds/ is a changing repository, instead of using the composite one can just use one of the immutable children (but they vanish after the next release):
https://download.eclipse.org/eclipse/updates/4.32-I-builds/I20240330-1800/

**Comment by kohlschuetter** [comment]
@HannesWell

Thanks for the tip! That could be another option. (It would be great though to know which build contains which commits — it's not obvious to me where to find that information)

I think there's a more of a general problem with switching from a stable release to bleeding edge development.

For example, I'm pretty sure my Eclipse workspace will become incompatible, so downgrading will suddenly become unnecessarily difficult. Then, there's potential unexpected fallout due to new bugs, etc.

This is all nothing new, not to me, and not in general (I have contributed to and survived living on "next major OS" releases for several years, and dogfooding is necessary for any serious project).

It's just that there are "regular" Eclipse developers that don't consider themselves developers _of_ Eclipse but simply developers _using_ Eclipse, and they should certainly be able to just stick to the regular stable channel.

As @merks commented [above]( https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-2030430446), I now understand that shipping a patch for a botched release isn't something trivial at the moment for the Eclipse IDE project. I wonder if that situation merits some attention from project management.

Any thoughts? (follow-up maybe on a separate issue?)

**Comment by HannesWell** [comment]
> It would be great though to know which build contains which commits — it's not obvious to me where to find that information

For each I-build the commit build of each sub-project is tagged with `I-<build-id>`. So you from the interesting commit you can just check the next one tagged (maybe even that commit) and look for the matching I-build.


> As @merks commented [above]( https://github.com/eclipse-platform/eclipse.platform.ui/issues/1674#issuecomment-2030430446), I now understand that shipping a patch for a botched release isn't something trivial at the moment for the Eclipse IDE project. I wonder if that situation merits some attention from project management.
> 
> Any thoughts? (follow-up maybe on a separate issue?)
> 

It's a somehow known issue but I don't think one that can be generally solved soon.
The release process is complex and although there is steady work to simplify it, it will certainly take time.
This is also a point discussed for example in
https://github.com/eclipse-pde/eclipse.pde/issues/1136#issuecomment-2033131726

But besides the man-power there are also technical difficulties for projects that are more or less a root of the decency tree like SWT. Due to the way for example features work a new version of SWT means that all features including it and the features including those features need new builds because versions are locked in features.
Otherwise the new version is not considered by you installation (which is usually based on features).
In unfortunate circumstances this could mean that many features in the entire SimRel would need to be rebuilt.
We are working on using features more 'precisely' but it is a long journey and not all projects in the SimRel are eager to go it.

**Comment by merks** [comment]
FYI. On my todo list is to investigate how best to allow EPP packages, as on the main download page, to update to any new platform build, including I-builds.  It’s annoying when a problem in a simrel milestone or a release cannot be updated until there is a new simrel milestone. And I think it’s an avoidable annoyance.
